subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildTriggerHandler.java

```diff
@@ -0,0 +1,279 @@
+
+import org.gradle.api.execution.internal.FileChangeListener;
+import org.gradle.api.execution.internal.TaskInputsListener;
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.FileCollectionStructureVisitor;
+import org.gradle.api.internal.file.FileTreeInternal;
+import org.gradle.api.internal.file.collections.FileSystemMirroringFileTree;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.execution.plan.BuildInputHierarchy;
+import org.gradle.initialization.BuildCancellationToken;
+import org.gradle.initialization.ContinuousExecutionGate;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.filewatch.PendingChangesListener;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.internal.watch.registry.FileWatcherRegistry;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.gradle.internal.filewatch.DefaultFileSystemChangeWaiterFactory.QUIET_PERIOD_SYSPROP;
+
+public class ContinuousBuildTriggerHandler implements FileChangeListener, TaskInputsListener {
+    private static final boolean IS_MAC_OSX = OperatingSystem.current().isMacOsX();
+
+    private final BuildInputHierarchy buildInputHierarchy;
+    private final PendingChangesListener pendingChangesListener;
+    private final BuildCancellationToken cancellationToken;
+    private final ContinuousExecutionGate continuousExecutionGate;
+    private final AtomicBoolean changeArrived = new AtomicBoolean(false);
+    private final CountDownLatch changeOrCancellationArrived = new CountDownLatch(1);
+    private final CountDownLatch cancellationArrived = new CountDownLatch(1);
+    private final FileEventCollector fileEventCollector = new FileEventCollector();
+    private final long quietPeriod;
+    private volatile long lastChangeAt = monotonicClockMillis();
+
+    public ContinuousBuildTriggerHandler(
+        BuildInputHierarchy buildInputHierarchy,
+        PendingChangesListener pendingChangesListener,
+        BuildCancellationToken cancellationToken,
+        ContinuousExecutionGate continuousExecutionGate
+    ) {
+        this.buildInputHierarchy = buildInputHierarchy;
+        this.pendingChangesListener = pendingChangesListener;
+        this.cancellationToken = cancellationToken;
+        this.continuousExecutionGate = continuousExecutionGate;
+        this.quietPeriod = Long.getLong(QUIET_PERIOD_SYSPROP, 250L);
+    }
+
+    public boolean hasAnyInputs() {
+        return !buildInputHierarchy.isEmpty();
+    }
+
+    void wait(Runnable notifier) {
```

Comment:
Should this be `public`? If we want to use default visibility, then `reportChanges()` should be that, too. I'd go for all `public` instead.

Reply:
I made it `public`.